File: Project2/Dotproduct.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/time.h>
#include <time.h>

#define CYCLE_TIME 100

int calculate_vector_int(int* data, int N, int dim);
int calculate_vector_short(short* data, int N, int dim);
int calculate_vector_byte(int8_t* data, int N, int dim);
double calculate_vector_double(double* data, int N, int dim);
float calculate_vector_float(float* data, int N, int dim);

long time_diff(struct timespec *start, struct timespec *end) {
    if (start && end) {
        return (end->tv_sec - start->tv_sec) * 1e9 + (end->tv_nsec - start->tv_nsec);
    } else {
        printf("wrong time interval parsing\n");
        return 0;
    }
}

int main() {
    const char* filename = "randomCase.txt";
    FILE* file = fopen(filename, "r");
    if (!file) {
        perror("can't open file");
        return 1;
    }

    int type, dim, N;
    if (fscanf(file, "%d %d %d", &type, &dim, &N) != 3) {
        printf("wrong file header\n");
        fclose(file);
        return 1;
    }

    void* data = NULL;
    long double sum = 0;
    struct timespec start, end;
    long avg_interval = 0;

    switch (type) {
        case 1: { // int
            data = malloc(N * dim * sizeof(int));
            int* flat_data = (int*)data;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < dim; j++) {
                    if (fscanf(file, "%d", &flat_data[i * dim + j]) != 1) {
                        printf("parse number failed\n");
                        goto cleanup;
                    }
                }
            }
            for (int i = 0; i < CYCLE_TIME; i++) {
                clock_gettime(CLOCK_MONOTONIC, &start);
                sum = calculate_vector_int(flat_data, N, dim);
                clock_gettime(CLOCK_MONOTONIC, &end);
                avg_interval += time_diff(&start, &end);
            }
            avg_interval /= CYCLE_TIME;
            break;
        }
        case 2: { // short
            data = malloc(N * dim * sizeof(short));
            short* flat_data = (short*)data;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < dim; j++) {
                    if (fscanf(file, "%hd", &flat_data[i * dim + j]) != 1) {
                        printf("parse number failed\n");
                        goto cleanup;
                    }
                }
            }
            for (int i = 0; i < CYCLE_TIME; i++) {
                clock_gettime(CLOCK_MONOTONIC, &start);
                sum = calculate_vector_short(flat_data, N, dim);
                clock_gettime(CLOCK_MONOTONIC, &end);
                avg_interval += time_diff(&start, &end);
            }
            avg_interval /= CYCLE_TIME;
            break;
        }
        case 3: { // byte (int8_t)
            data = malloc(N * dim * sizeof(int8_t));
            int8_t* flat_data = (int8_t*)data;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < dim; j++) {
                    int tmp;
                    if (fscanf(file, "%d", &tmp) != 1) {
                        printf("parse number failed\n");
                        goto cleanup;
                    }
                    flat_data[i * dim + j] = (int8_t)tmp;
                }
            }
            for (int i = 0; i < CYCLE_TIME; i++) {
                clock_gettime(CLOCK_MONOTONIC, &start);
                sum = calculate_vector_byte(flat_data, N, dim);
                clock_gettime(CLOCK_MONOTONIC, &end);
                avg_interval += time_diff(&start, &end);
            }
            avg_interval /= CYCLE_TIME;
            break;
        }
        case 4:{ // double
            data = malloc(N * dim * sizeof(double));
            double* flat_data = (double*)data;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < dim; j++) {
                    if (fscanf(file, "%lf", &flat_data[i * dim + j]) != 1) {
                        printf("parse number failed\n");
                        goto cleanup;
                    }
                }
            }
            for (int i = 0; i < CYCLE_TIME; i++) {
                clock_gettime(CLOCK_MONOTONIC, &start);
                sum = calculate_vector_double(flat_data, N, dim);
                clock_gettime(CLOCK_MONOTONIC, &end);
                avg_interval += time_diff(&start, &end);
            }
            avg_interval /= CYCLE_TIME;
            break;
        }
        case 5: { // float
            data = malloc(N * dim * sizeof(float));
            float* flat_data = (float*)data;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < dim; j++) {
                    if (fscanf(file, "%f", &flat_data[i * dim + j]) != 1) {
                        printf("parse number failed\n");
                        goto cleanup;
                    }
                }
            }
            for (int i = 0; i < CYCLE_TIME; i++) {
                clock_gettime(CLOCK_MONOTONIC, &start);
                sum = calculate_vector_float(flat_data, N, dim);
                clock_gettime(CLOCK_MONOTONIC, &end);
                avg_interval += time_diff(&start, &end);
            }
            avg_interval /= CYCLE_TIME;
            break;
        }
        default:
            printf("unsupported type: %d\n", type);
            goto cleanup;
    }

    printf("sum = %Lf\n", sum);
    printf("Calculations finished in avg %ld nano second\n", avg_interval);
    printf("Calculations finished in avg %.6f millisecond\n", avg_interval / 1e6);

cleanup:
    if (data) free(data);
    fclose(file);
    return 0;
}

int calculate_vector_int(int* data, int N, int dim) {
    int sum = 0;
    if (!data) {
        printf("empty data passing\n");
        return sum;
    }
    for (int i = 1; 2 * i <= N; i++) {
        int offset1 = (2 * i - 2) * dim;//0 2 4
        int offset2 = (2 * i - 1) * dim;//1 3 5
        for (int j = 0; j < dim; j++) {
            sum += data[offset1 + j] * data[offset2 + j];
        }
    }
    return sum;
}

int calculate_vector_short(short* data, int N, int dim) {
    int sum = 0;
    if (!data) {
        printf("empty data passing\n");
        return sum;
    }
    for (int i = 1; 2 * i <= N; i++) {
        int offset1 = (2 * i - 2) * dim;
        int offset2 = (2 * i - 1) * dim;
        for (int j = 0; j < dim; j++) {
            sum += data[offset1 + j] * data[offset2 + j];
        }
    }
    return sum;
}

int calculate_vector_byte(int8_t* data, int N, int dim) {
    int sum = 0;
    if (!data) {
        printf("empty data passing\n");
        return sum;
    }
    for (int i = 1; 2 * i <= N; i++) {
        int offset1 = (2 * i - 2) * dim;
        int offset2 = (2 * i - 1) * dim;
        for (int j = 0; j < dim; j++) {
            sum += data[offset1 + j] * data[offset2 + j];
        }
    }
    return sum;
}

double calculate_vector_double(double* data, int N, int dim) {
    double sum = 0;
    if (!data) {
        printf("empty data passing\n");
        return sum;
    }
    for (int i = 1; 2 * i <= N; i++) {
        int offset1 = (2 * i - 2) * dim;
        int offset2 = (2 * i - 1) * dim;
        for (int j = 0; j < dim; j++) {
            sum += data[offset1 + j] * data[offset2 + j];
        }
    }
    return sum;
}

float calculate_vector_float(float* data, int N, int dim) {
    float sum = 0;
    if (!data) {
        printf("empty data passing\n");
        return sum;
    }
    for (int i = 1; 2 * i <= N; i++) {
        int offset1 = (2 * i - 2) * dim;
        int offset2 = (2 * i - 1) * dim;
        for (int j = 0; j < dim; j++) {
            sum += data[offset1 + j] * data[offset2 + j];
        }
    }
    return sum;
}

File: Project2/Dotproduct.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.StringTokenizer;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.StringTokenizer;

public class JavaVecMulti {

    static final int CYCLE_TIME = 100;

    public static void main(String[] args) throws IOException {
        String inputFile = "randomCase.txt";
        QReader sc = new QReader(new File(inputFile));
        int type = sc.nextInt();
        int dim = sc.nextInt();
        int N = sc.nextInt();

        double sum = 0;
        long totalTime = 0;

        for (int i = 0; i < 10000; i++) System.nanoTime();

        switch (type) {
            case 1 -> { // int
                int[] intData = new int[N * dim];
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < dim; j++) {
                        intData[i * dim + j] = sc.nextInt();
                    }
                }
                for (int i = 0; i < CYCLE_TIME; i++) {
                    long start = System.nanoTime();
                    sum = calculateVectorInt(intData, N, dim);
                    totalTime += System.nanoTime() - start;
                }
            }
            case 2 -> { // short
                short[] shortData = new short[N * dim];
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < dim; j++) {
                        shortData[i * dim + j] = sc.nextShort();
                    }
                }
                for (int i = 0; i < CYCLE_TIME; i++) {
                    long start = System.nanoTime();
                    sum = calculateVectorShort(shortData, N, dim);
                    totalTime += System.nanoTime() - start;
                }
            }
            case 3 -> { // byte
                byte[] byteData = new byte[N * dim];
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < dim; j++) {
                        byteData[i * dim + j] = sc.nextByte();
                    }
                }
                for (int i = 0; i < CYCLE_TIME; i++) {
                    long start = System.nanoTime();
                    sum = calculateVectorByte(byteData, N, dim);
                    totalTime += System.nanoTime() - start;
                }
            }
            case 4 -> { // double
                double[] doubleData = new double[N * dim];
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < dim; j++) {
                        doubleData[i * dim + j] = sc.nextDouble();
                    }
                }
                for (int i = 0; i < CYCLE_TIME; i++) {
                    long start = System.nanoTime();
                    sum = calculateVectorDouble(doubleData, N, dim);
                    totalTime += System.nanoTime() - start;
                }
            }
            case 5 -> { // float
                float[] floatData = new float[N * dim];
                for (int i = 0; i < N; i++) {
                    for (int j = 0; j < dim; j++) {
                        floatData[i * dim + j] = sc.nextFloat();
                    }
                }
                for (int i = 0; i < CYCLE_TIME; i++) {
                    long start = System.nanoTime();
                    sum = calculateVectorFloat(floatData, N, dim);
                    totalTime += System.nanoTime() - start;
                }
            }
            default -> throw new IllegalArgumentException("Invalid type: " + type);
        }

        long avgTime = totalTime / CYCLE_TIME;
        System.out.printf("sum = %f\n", sum);
        System.out.println("Calculations finished in avg "+avgTime+" nano second");
        System.out.println("Calculations finished in avg "+avgTime/1e6+" millisecond");
    }


    public static int calculateVectorInt(int[] flatData, int N, int dim) {
        int sum = 0;
        for (int i = 1; 2 * i <= N; i++) {
            int offset1 = (2 * i - 2) * dim;
            int offset2 = (2 * i - 1) * dim;
            for (int j = 0; j < dim; j++) {
                sum += flatData[offset1 + j] * flatData[offset2 + j];
            }
        }
        return sum;
    }

    public static int calculateVectorShort(short[] data, int N, int dim) {
        int sum = 0;
        for (int i = 1; 2 * i <= N; i++) {
            int p1 = (2 * i - 2) * dim, p2 = (2 * i - 1) * dim;
            for (int j = 0; j < dim; j++) {
                sum += data[p1 + j] * data[p2 + j];
            }
        }
        return sum;
    }

    public static int calculateVectorByte(byte[] data, int N, int dim) {
        int sum = 0;
        for (int i = 1; 2 * i <= N; i++) {
            int p1 = (2 * i - 2) * dim, p2 = (2 * i - 1) * dim;
            for (int j = 0; j < dim; j++) {
                sum += data[p1 + j] * data[p2 + j];
            }
        }
        return sum;
    }

    public static float calculateVectorFloat(float[] data, int N, int dim) {
        float sum = 0;
        for (int i = 1; 2 * i <= N; i++) {
            int p1 = (2 * i - 2) * dim, p2 = (2 * i - 1) * dim;
            for (int j = 0; j < dim; j++) {
                sum += data[p1 + j] * data[p2 + j];
            }
        }
        return sum;
    }

    public static double calculateVectorDouble(double[] flatData, int N, int dim) {
        double sum = 0;
        for (int i = 1; 2 * i <= N; i++) {
            int offset1 = (2 * i - 2) * dim;
            int offset2 = (2 * i - 1) * dim;
            for (int j = 0; j < dim; j++) {
                sum += flatData[offset1 + j] * flatData[offset2 + j];
            }
        }
        return sum;
    }
}


class QReader {
    private BufferedReader reader;
    private StringTokenizer tokenizer;

    public QReader(){
        reader = new BufferedReader(new InputStreamReader(System.in));
        tokenizer = new StringTokenizer("");
    }

    public QReader(File file){
        try {
            if (file.isFile() && file.exists()) {
                InputStreamReader inputReader = new InputStreamReader(Files.newInputStream(file.toPath()), StandardCharsets.UTF_8);
                reader = new BufferedReader(inputReader);
            }
        }catch (IOException e){
            System.err.println("File not found: " + e.getMessage());
        }
        tokenizer = new StringTokenizer("");
    }

    private String innerNextLine() {
        try {
            return reader.readLine();
        } catch (IOException e) {
            return null;
        }
    }

    public boolean hasNext() {
        while (!tokenizer.hasMoreTokens()) {
            String nextLine = innerNextLine();
            if (nextLine == null) {
                return false;
            }
            tokenizer = new StringTokenizer(nextLine);
        }
        return true;
    }

    public String nextLine() {
        tokenizer = new StringTokenizer("");
        return innerNextLine();
    }

    public String next() {
        hasNext();
        return tokenizer.nextToken();
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

    public short nextShort() {
        return Short.parseShort(next());
    }

    public byte nextByte() {
        return Byte.parseByte(next());
    }


    public float nextFloat() {
        return Float.parseFloat(next());
    }

    public double nextDouble() {
        return Double.parseDouble(next());
    }

    public long nextLong() {
        return Long.parseLong(next());
    }
}

class QWriter implements Closeable {
    private BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

    public void print(Object object) {
        try {
            writer.write(object.toString());
        } catch (IOException e) {
            return;
        }
    }

    public void println(Object object) {
        try {
            writer.write(object.toString());
            writer.write("\n");
        } catch (IOException e) {
            return;
        }
    }

    public void println() {
        try {
            writer.write("\n");
        } catch (IOException e) {
            return;
        }
    }

    @Override
    public void close() {
        try {
            writer.close();
        } catch (IOException e) {
            return;
        }
    }
}
